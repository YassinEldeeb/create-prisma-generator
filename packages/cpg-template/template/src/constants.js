export const GENERATOR_NAME = '$PACKAGE_NAME'
