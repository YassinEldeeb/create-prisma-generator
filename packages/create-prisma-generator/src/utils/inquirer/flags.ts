export const flags = { skipPrismaNamingConventionFlag: '--skip-check' }
