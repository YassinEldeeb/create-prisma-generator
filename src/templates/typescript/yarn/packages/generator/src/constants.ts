export const GENERATOR_NAME = 'test_package'
