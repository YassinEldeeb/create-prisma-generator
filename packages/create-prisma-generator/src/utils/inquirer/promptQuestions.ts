import inquirer from 'inquirer'
import { Answers } from '../../types/answers'
import { filterValue } from './filters/value'
import { validateGeneratorName } from './validations/generatorName'

// ANCHOR[id=questions]
export const questions: inquirer.QuestionCollection<any>[] = [
  // ANCHOR[id=Q1-generatorName]
  {
    type: 'input',
    name: 'generatorName',
    message: "what's your generator name",
    // Validate Package Name
    validate(pkgName: string) {
      return validateGeneratorName(pkgName)
    },
  },
  // ANCHOR[id=Q2-usingTypescript]
  {
    type: 'confirm',
    name: 'typescript',
    message: 'do you want to use Typescript',
  },
  // ANCHOR[id=Q3-selectPkgManager]
  {
    type: 'list',
    name: 'packageManager',
    message: 'which package manager do you want to use',
    choices: ['yarn', 'pnpm', 'npm'],
    default: 'pnpm',
  },
  // ANCHOR[id=Q4-usingGithubActions]
  {
    type: 'confirm',
    name: 'githubActions',
    message: 'automate publishing the generator with Github Actions',
  },
  // ANCHOR[id=Q5-enableSemanticRelease]
  {
    type: 'confirm',
    name: 'semanticRelease',
    message: '(Github Actions) setup automatic semantic release',
    when: (answers: Answers) => {
      return answers.githubActions
    },
  },
  // ANCHOR[id=Q5-setupWorkspaceWithUsage]
  {
    type: 'confirm',
    name: 'usageTemplate',
    message: 'setup workspace for testing the generator',
  },
]

export const promptQuestions = async (): Promise<Answers> => {
  const answers = (await inquirer.prompt(questions)) as Answers

  Object.keys(answers).forEach((e) => {
    const value = answers[e as keyof typeof answers]
    ;(answers as any)[e] = filterValue(value)
  })

  return answers
}
