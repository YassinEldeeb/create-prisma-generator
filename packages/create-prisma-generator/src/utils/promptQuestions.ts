import inquirer from 'inquirer'
import { Answers } from '../types/answers'

export const promptQuestions = async (): Promise<Answers> => {
  const questions: inquirer.QuestionCollection<any> = [
    {
      type: 'input',
      name: 'generatorName',
      message: "What's your generator name",
    },
    {
      type: 'confirm',
      name: 'typescript',
      message: 'Do you want to use Typescript',
    },
    {
      type: 'list',
      name: 'packageManager',
      message: 'Which package manager do you want to use',
      choices: ['npm', 'yarn', 'pnpm'],
      default: false,
    },
    {
      type: 'confirm',
      name: 'githubAction',
      message: 'Using github? Can I automate some stuff for you',
    },
  ]
  const answers = (await inquirer.prompt(questions)) as Answers

  return answers
}
