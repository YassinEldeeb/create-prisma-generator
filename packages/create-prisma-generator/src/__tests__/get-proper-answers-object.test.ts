import { MockSTDIN, stdin } from 'mock-stdin'
import { promptQuestions } from '../utils/inquirer/promptQuestions'
import { skipQuestions } from './__helpers__/skipQuestions'
import { answer } from './__helpers__/answer'
import { validGenName } from './constants/valid-prisma-gen-name'

// Mock stdin to send keystrokes to the CLI
let io: MockSTDIN
beforeEach(() => (io = stdin()))
afterEach(() => io.restore())

// See the Questions to know how to modify these samples
// LINK ..\utils\promptQuestions.ts#questions

let genName = validGenName
const sampleAnswers = {
  async sample1() {
    // LINK ..\utils\promptQuestions.ts#Q1-generatorName
    await answer(io, { text: `  ${genName}    ` })

    // Skip the rest of the questions
    await skipQuestions(-1, io)
  },
  async sample2() {
    // LINK ..\utils\promptQuestions.ts#Q1-generatorName
    await answer(io, { text: genName })

    // Skip the rest of the questions
    await skipQuestions(-1, io, true)
  },
  async sample3() {
    // LINK ..\utils\promptQuestions.ts#Q1-generatorName
    await answer(io, { text: genName })

    // LINK ..\utils\promptQuestions.ts#Q2-usingTypescript
    await answer(io, { text: 'No' })

    // LINK ..\utils\promptQuestions.ts#Q3-selectPkgManager
    await answer(io, { keys: ['down'] })

    // LINK ..\utils\promptQuestions.ts#Q4-usingGithubActions
    await answer(io, { text: 'No' })

    // Skip the rest of the questions
    await skipQuestions(-1, io)
  },
  async sample4() {
    // LINK ..\utils\promptQuestions.ts#Q1-generatorName
    await answer(io, { text: 'my-gen --skip-check' })

    // Skip the rest of the questions
    await skipQuestions(2, io)
    await skipQuestions(-1, io, true)
  },
}

Object.keys(sampleAnswers).map((answer) => {
  test(`should get proper answers object when answering using ${answer}`, async () => {
    setTimeout(
      () => sampleAnswers[answer as keyof typeof sampleAnswers]().then(),
      5,
    )
    const answers = await promptQuestions()

    expect(answers).toMatchSnapshot()
  })
})
