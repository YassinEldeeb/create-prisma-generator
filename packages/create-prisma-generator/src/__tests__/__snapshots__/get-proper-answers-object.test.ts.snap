// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get proper answers object when answering using sample1 1`] = `
Object {
  "generatorName": "prisma-generator-data-graph",
  "githubActions": true,
  "packageManager": "pnpm",
  "semanticRelease": true,
  "typescript": true,
  "usageTemplate": true,
}
`;

exports[`should get proper answers object when answering using sample2 1`] = `
Object {
  "generatorName": "prisma-generator-data-graph",
  "githubActions": false,
  "packageManager": "pnpm",
  "typescript": false,
  "usageTemplate": false,
}
`;

exports[`should get proper answers object when answering using sample3 1`] = `
Object {
  "generatorName": "prisma-generator-data-graph",
  "githubActions": false,
  "packageManager": "npm",
  "typescript": false,
  "usageTemplate": true,
}
`;

exports[`should get proper answers object when answering using sample4 1`] = `
Object {
  "generatorName": "my-gen",
  "githubActions": false,
  "packageManager": "pnpm",
  "typescript": true,
  "usageTemplate": false,
}
`;
